# Laboratorio 2

Bienvenido al segundo laboratorio del curso Análisis de Datos en Python. 

Escribe el código correspondiente para contestar cada una de las preguntas. Fíjate muy bien en el nombre de las variables, ya que de lo contrario, tu trabajo no podrá ser evaluado de manera correcta. 

Debes ir ejecutando las celdas de codigo en el orden que aparecen, para evitar tener errores. No elimines celdas!! 

Este trabajo tiene un límite de tiempo que puedes visualizar al inicio de esta pantalla. 

### Pregunta 1
¿Cuál es el promedio y desviación estándar de la cantidad de basura diaria generada en cada calle? 

Indicaciones: 
- Guarda los datos en un dataframe, bajo la variable "basuraxcalle"
- El dataframe debe tener 3 columnas: "Calle", "Cantidad promedio [kilos de basura]" y "Desv est [kilos de basura]" 

In [29]:
import pandas as pd
import numpy as np

ciudad = pd.read_excel('Datos.xlsx',sheet_name = 'Ciudad')
basura = pd.read_excel('Datos.xlsx',sheet_name = 'Histórico de basura')

basuraxcalle= None
import pandas as pd

# cambio de nombre a "Calle" en cuidad
ciudad = ciudad.rename(columns={'Se considera que la ciudad comienza en la esquina inferior izquierda de la cuadrícula que la representa, y entonces: X=0 e Y=0': 'Calle'})

# Union de dataframes
data = pd.merge(basura, ciudad, on='Calle')

# Agrupar los datos por calle y calcula el promedio y la desviación estándar
basuraxcalle = data.groupby(['Calle', 'Coordenada (X,Y) donde comienza la calle', 'Coordenada (X,Y) donde termina la calle'])['Cantidad [kilos de basura]'].agg(['mean', 'std']).reset_index()
basuraxcalle.columns = ['Calle', 'Coordenada Inicio', 'Coordenada Final', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]']

# Eliminar columnas "Coordenada Inicio" y "Coordenada Final"
basuraxcalle = basuraxcalle.drop(['Coordenada Inicio', 'Coordenada Final'], axis=1)

basuraxcalle



Calle  Cantidad promedio [kilos de basura]  \
0        Calle 1                             2.982143   
1       Calle 10                             3.192857   
2      Calle 100                             2.664286   
3     Calle 1000                             2.742857   
4     Calle 1001                             2.896429   
...          ...                                  ...   
3235   Calle 995                             3.032143   
3236   Calle 996                             2.203571   
3237   Calle 997                             2.964286   
3238   Calle 998                             2.978571   
3239   Calle 999                             2.867857   

      Desv est [kilos de basura]  
0                       1.209240  
1                       1.279447  
2                       1.199449  
3                       1.211191  
4                       1.093578  
...                          ...  
3235                    1.172858  
3236                    0.850918  
3237                    1.126520  
3238                    1.203500  
3239                    0.990717  

[3240 rows x 3 columns]

Revisemos que el dataframe haya quedado bien creado, la siguiente celda no tiene puntaje, pero te servirá para que puedas revisar que hayas creado bien tu dataframe 

In [30]:
assert list(basuraxcalle.columns) == ['Calle', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]'], "Los nombres de las columnas son incorrectos"

In [31]:
basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]

Calle  Cantidad promedio [kilos de basura]  Desv est [kilos de basura]
0  Calle 1                             2.982143                     1.20924

El resultado para la Calle 1 deberia ser


|Calle   | Cantidad promedio [kilos de basura]	  | Desv est [kilos de basura]  |
|---|---|---|
|Calle 1  | 2.982143  | 1.20924  |

No es necesario que redondees el numero de promedio o desviacion estandar, ni que hagas ninguna operación con el: guardalo en el dataframe tal cual lo entregue el calculo. 

Revisemos que hayas calculado bien el promedio y desviacion para la calle 1:

In [32]:
assert np.round(basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]["Cantidad promedio [kilos de basura]"].values,2) == [[2.98]], "Fallo en el promedio"
assert np.round(basuraxcalle[basuraxcalle['Calle'] == "Calle 1"]["Desv est [kilos de basura]"].values,2) == [[1.21]], "Fallo en la desviacion estandar"

La siguiente celda hace tests ocultos, en el mismo formato del anterior pero para otras calles seleccionadas al azar! No modifiques, muevas ni elimines la celda

### Pregunta 2

Se define como **calle “intensa”** a aquellas calles cuyo promedio diario de basura generada es **estrictamente mayor a 2,92 kilos por día**.

¿Cuántas de las calles serían consideradas intensas (en función de los datos históricos de 28 días)?

Guarda la cantidad de calles intensas en la variable c_intensas, en formato int. 

NO DEBES REDONDEAR NI APROXIMAR EL PROMEDIO PARA HACER LA CLASIFICACION!!

**Hint: usa el dataframe que construiste para la pregunta anterior!**

In [34]:
c_intensas=0

# Filtrar las calles intensas
calles_intensas = basuraxcalle[basuraxcalle['Cantidad promedio [kilos de basura]'] > 2.92]

# numero de calles intensas
c_intensas = len(calles_intensas)

# Muestra la cantidad de calles intensas
print(f"La cantidad de calles intensas es: {c_intensas}")

La cantidad de calles intensas es: 1657


La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 3

En el dataframe `ciudad` hay datos de las coordenadas donde comienzan y terminan las calles. 

La comuna comienza en la coordenada 0 y termina en la coordenada 40, tanto para X como para Y. Si se divide territorialmente la comuna de la siguiente manera:

#### a. En X, desde la coordenada 0 a la 22 (inclusive) se define Oeste (O)
#### b. En X, desde la coordenada 23 a la 40 (inclusive) se define Este (E)
#### c. En Y, desde la coordenada 0 a la 27 (inclusive) se define Norte (N)
#### d. En Y, desde la coordenada 28 a la 40 (inclusive) se define Sur (S)

Considerando esta división territorial, las calles de la comuna se clasifican como zonas: N-E, N-O, S-E, S-O. 

Puede suceder que alguna calle comience en una zona y termine en otra. Veamos un ejemplo!

Primero renombremos las columnas

In [35]:
ciudad = ciudad.rename(columns = {'Se considera que la ciudad comienza en la esquina inferior izquierda de la cuadrícula que la representa, y entonces: X=0 e Y=0':'Calle',
          'Coordenada (X,Y) donde comienza la calle':'Coordenada inicial',
       'Coordenada (X,Y) donde termina la calle':'Coordenada final'})

Ahora veamos por ejemplo la calle 28

In [36]:
ciudad[ciudad["Calle"] == "Calle 28"]

Calle Coordenada inicial Coordenada final  Unnamed: 3 Unnamed: 4
27  Calle 28             (0,27)           (0,28)         NaN        NaN

Esta calle comienza en la coordenada (0,27), pero termina en (0,28). Esto quiere decir, que empieza en la zona N-O, pero termina en la zona S-O. Esta es una clase que no se puede clasificar!!

A continuacion, guarda en la variable `sin_clasificar` la cantidad de calles que no pueden ser clasificadas en una zona, es decir, que comiencen en una y terminen en otra. El formato debe ser int

In [38]:
sin_clasificar=None
# your code here

# Define las condiciones para clasificar las calles en las zonas
condiciones = [
    (ciudad['Coordenada inicial'].str[0] <= 22) & (ciudad['Coordenada final'].str[0] <= 22) & (ciudad['Coordenada inicial'].str[1] <= 27) & (ciudad['Coordenada final'].str[1] <= 27),
    (ciudad['Coordenada inicial'].str[0] <= 22) & (ciudad['Coordenada final'].str[0] <= 22) & (ciudad['Coordenada inicial'].str[1] > 27) & (ciudad['Coordenada final'].str[1] > 27),
    (ciudad['Coordenada inicial'].str[0] > 22) & (ciudad['Coordenada final'].str[0] > 22) & (ciudad['Coordenada inicial'].str[1] <= 27) & (ciudad['Coordenada final'].str[1] <= 27),
    (ciudad['Coordenada inicial'].str[0] > 22) & (ciudad['Coordenada final'].str[0] > 22) & (ciudad['Coordenada inicial'].str[1] > 27) & (ciudad['Coordenada final'].str[1] > 27)
]

# Define las opciones para las zonas
zonas = ['N-O', 'N-E', 'S-O', 'S-E']

# Aplica las condiciones y asigna las zonas correspondientes
ciudad['Zona'] = np.select(condiciones, zonas, default='Otro')

# Combina la información de la zona en el DataFrame basuraxcalle
basuraxcalle2 = pd.merge(basuraxcalle, ciudad[['Calle', 'Zona']], on='Calle')

# Muestra el DataFrame resultante con la clasificación de las calles en las zonas
basuraxcalle2


TypeError: '<=' not supported between instances of 'str' and 'int'

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 4
Considerando la clasificación anterior,
¿cuántas calles quedan clasificadas en cada zona?

Para contestar, genera un dataframe que tenga 2 columnas: Clasificacion y Calle. Dicho df debe ser almacenado en la variable "callesxzona".

Los valores para clasificacion son: "N-E", "N-O", "No clasificado", "S-E", "S-O"

Ejemplo del dataframe: 

|Clasificacion|Calle|
|---|---|
|N-E| 0|
|N-O| 0|
|No clasificado| 0|
|S-E| 0|
|S-O| 0|



In [ ]:
callesxzona=None
# your code here
raise NotImplementedError

Revisa que hayas creado bien el dataframe!

In [ ]:
print(list(callesxzona.columns) == ['Clasificacion', 'Calle'])

clasificaciones = list(callesxzona["Clasificacion"].values)
clasificaciones.sort()

print(clasificaciones == ['N-E', 'N-O', 'No clasificado', 'S-E', 'S-O'])

Comenzamos con un test visible!

In [ ]:
assert callesxzona[callesxzona['Clasificacion'] == 'N-E']["Calle"].values == [[962]]

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda

### Pregunta 5 
¿Cuál es el promedio y desviación estándar de la cantidad de kilos de basura generada por cada zona?

Excluya las calles que quedan sin clasificar. 

Guarde su respuesta en un dataframe con la siguiente estructura: 

|Zona   | Cantidad promedio [kilos de basura]	  | Desv est [kilos de basura]  |
|---|---|---|
|N-O  | 0  | 0  |

Guarde el dataframe en la variable `zona_stats`

In [ ]:
zona_stats=None
# your code here
raise NotImplementedError

Revisemos el formato

In [ ]:
zona_stats

In [ ]:
print(list(zona_stats.columns) == ['Zona', 'Cantidad promedio [kilos de basura]', 'Desv est [kilos de basura]'])

cols = list(zona_stats['Zona'].values)
cols.sort()

print(cols == ['N-E', 'N-O', 'S-E', 'S-O'])

Comenzamos con un test visible:

In [ ]:
ans_NE = np.round(zona_stats[zona_stats["Zona"]=='N-E'][["Cantidad promedio [kilos de basura]", "Desv est [kilos de basura]"]].values,3)

assert (ans_NE== [[2.926, 1.122]]).all()

La siguiente celda hace tests ocultos. No modifiques, muevas ni elimines la celda